server/errorhandler_test.go

```diff
@@ -0,0 +1,81 @@
+package server
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/containous/traefik/middlewares"
+)
+
+type timeoutError struct{}
+
+func (e *timeoutError) Error() string   { return "i/o timeout" }
+func (e *timeoutError) Timeout() bool   { return true }
+func (e *timeoutError) Temporary() bool { return true }
+
+func TestServeHTTP(t *testing.T) {
+	tests := []struct {
+		name               string
+		err                error
+		wantHTTPStatus     int
+		wantNetErrRecorded bool
+	}{
+		{
+			name:               "net.Error",
+			err:                net.UnknownNetworkError("any network error"),
+			wantHTTPStatus:     http.StatusBadGateway,
+			wantNetErrRecorded: true,
+		},
+		{
+			name:               "net.Error with Timeout",
+			err:                &timeoutError{},
+			wantHTTPStatus:     http.StatusGatewayTimeout,
+			wantNetErrRecorded: true,
+		},
+		{
+			name:               "io.EOF",
+			err:                io.EOF,
+			wantHTTPStatus:     http.StatusBadGateway,
+			wantNetErrRecorded: true,
+		},
+		{
+			name:               "custom error",
+			err:                errors.New("any error"),
+			wantHTTPStatus:     http.StatusInternalServerError,
+			wantNetErrRecorded: false,
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			recorder := httptest.NewRecorder()
+
+			netErrorOccured := false
+			req, err := http.NewRequest("GET", "http://localhost:3000/any", nil)
+			if err != nil {
+				panic(fmt.Sprintf("NewRequest(): %s", err))
```

Comment:
There is a util function in `httptest.NewRequest` that will internally panic if the creation of the `NewRequest` fails. I think this would be good to use for the tests.